Fixes update service endpoints in place (resolves recent vmi networking test failures)

davidvossel opened the pull request:
We've recently noticed some odd behavior in our Service endpoints. This behavior was exasperated further by a change to our operator that caused resources to be acted upon more frequently.

The issue is that ClusterIP service endpoints are not Patched in place. Instead they are deleted and re-created. This causes downtime to our API and can cause transient issues in our functional testsuite. 

This issue appeared releated to the recent CustomizePatches PR #3612 but was in reality completely unrelated. That PR did however introduce a naming collision with our api. KubeVirt v1.Patch and Kubernetes v1.Patch both exist. I fixed this by simply renaming our struct so it doesn't collide. 

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Fixes issue with Serivce endpoints not being updated properly in place during KubeVirt updates. 
```

xpivarc commented on staging/src/kubevirt.io/client-go/api/v1/types.go:
```diff
@@ -1172,11 +1172,11 @@ type KubeVirtSpec struct {
 
 // +k8s:openapi-gen=true
-type Patch struct {
+type CustomizeComponentsPatch struct {
```
I am all for this. But we should consider to improve our generating of openapi so this problem doesn't come up in the future.

davidvossel replied:
yes. either improve our openapi generation logic, or create some sort of test case that caches when we collide with a Kubernetes api field.

dhiller replied:
I'd prefer the latter.

xpivarc commented on pkg/virt-operator/install-strategy/create.go:
```diff
@@ -1415,10 +1419,18 @@ func generateServicePatch(kv *v1.KubeVirt,
-		// It is okay if the cached service has a ClusterIP and the target one is empty. We just
-		// ensure the cached ClusterIP is carried over in the Patch.
-		deleteAndReplace = true
+		if service.Spec.ClusterIP == "" {
```
Why do we rely on ClusterIP?

davidvossel replied:
Service resources are a bit strange.  If you set the ClusterIP == "", then the k8s backend automatically assigns an IP address. The weird thing here for us is that the k8s backend actually mutates the service's spec with that new ClusterIP field.

So, the install strategy's desired service has `ClusterIP == ""` and the cached service has `ClusterIP == <some dynamic IP>`

Typically for our other resources we'd simply stomp on any changes that differ in the spec from our desired spec, versus what we see cached. However for a ClusterIP service, we actually need to preserve the cached ClusterIP field and merge it with the install strategies desired service spec.

Basically, cluster ip is something that we set to "" and magically turns into an IP later on. This inconsistency between our desired spec and what it gets mutated into requires some special handling.

xpivarc replied:
Thanks, that makes absolute sense :+1:

jean-edouard commented:
/approve
Good catch(es)! Looks like there are more instances of `Patch` to fix (seen in travis-ci: `pkg/virt-operator/install-strategy/patches.go:21:12: undefined: "kubevirt.io/client-go/api/v1".Patch`).
Approving ahead of that getting addressed (as well as the other comments above) , since the general idea makes sense and we need this asap.

davidvossel commented:
> Good catch(es)! Looks like there are more instances of Patch to fix (seen in travis-ci: pkg/virt-operator/install-strategy/patches.go:21:12: undefined: "kubevirt.io/client-go/api/v1".Patch).

yup, fixed. I added some unit tests as well for the ClusterIP logic.

xpivarc reviewed (commented):
There are some leftovers, otherwise
\lgtm

oshoval commented on api/openapi-spec/swagger.json:
```diff
@@ -6632,12 +6632,29 @@
      }
     }
    },
+   "v1.CustomizeComponentsPatch": {
```
Please revert this file, unless you did it by purpose of this PR.
Reason is there is a known issue of flakiness generating swagger.json

davidvossel replied:
This pr is meant to address all the current dev issues we're encountering. This fixes the flaky swagger generation logic... at least for now. A better fix is welcome in the future, this is a simply way to get unblocked though that doesn't impact anything reallly.

oshoval commented on pkg/virt-operator/install-strategy/create_test.go:
```diff
@@ -278,6 +279,83 @@ var _ = Describe("Create", func() {
+
+			cachedService := &corev1.Service{}
+			cachedService.Spec.Type = corev1.ServiceTypeClusterIP
+			cachedService.Spec.ClusterIP = "10.10.10.10"
```
do we need to add ipv6 test here ? or its not important ?

davidvossel replied:
not important. i could set this to anything that's != "" for the purposes here.

davidvossel commented:
/retest

xpivarc commented:
/lgtm

mhenriks commented:
/lgtm

xpivarc commented:
/test pull-kubevirt-e2e-k8s-1.19

mhenriks commented:
/retest

dhiller commented:
/retest
